src/main/javassist/bytecode/AttributeInfo.java

```diff
@@ -79,65 +80,58 @@ static AttributeInfo read(ConstPool cp, DataInputStream in)
-                else if (nameStr.equals(ConstantAttribute.tag))
-                    return new ConstantAttribute(cp, name, in);
-                else if (nameStr.equals(DeprecatedAttribute.tag))
-                    return new DeprecatedAttribute(cp, name, in);
-            }
-            else {
-                if (nameStr.equals(EnclosingMethodAttribute.tag))
-                    return new EnclosingMethodAttribute(cp, name, in);
-                else if (nameStr.equals(ExceptionsAttribute.tag))
-                    return new ExceptionsAttribute(cp, name, in);
-                else if (nameStr.equals(InnerClassesAttribute.tag))
-                    return new InnerClassesAttribute(cp, name, in);
-                else if (nameStr.equals(LineNumberAttribute.tag))
-                    return new LineNumberAttribute(cp, name, in);
-                else if (nameStr.equals(LocalVariableAttribute.tag))
-                    return new LocalVariableAttribute(cp, name, in);
-                else if (nameStr.equals(LocalVariableTypeAttribute.tag))
-                    return new LocalVariableTypeAttribute(cp, name, in);
-            }
-        }
-        else {
-            if (first < 'S') {
-                /* Note that the names of Annotations attributes begin with 'R'. 
-                 */
-                if (nameStr.equals(MethodParametersAttribute.tag))
-                    return new MethodParametersAttribute(cp, name, in);
-                else if (nameStr.equals(AnnotationsAttribute.visibleTag)
-                         || nameStr.equals(AnnotationsAttribute.invisibleTag)) {
-                    // RuntimeVisibleAnnotations or RuntimeInvisibleAnnotations
-                    return new AnnotationsAttribute(cp, name, in);
-                }
-                else if (nameStr.equals(ParameterAnnotationsAttribute.visibleTag)
-                         || nameStr.equals(ParameterAnnotationsAttribute.invisibleTag))
-                    return new ParameterAnnotationsAttribute(cp, name, in);
-                else if (nameStr.equals(TypeAnnotationsAttribute.visibleTag)
-                         || nameStr.equals(TypeAnnotationsAttribute.invisibleTag))
-                    return new TypeAnnotationsAttribute(cp, name, in);
-            }
-            else {
-                if (nameStr.equals(SignatureAttribute.tag))
-                    return new SignatureAttribute(cp, name, in);
-                else if (nameStr.equals(SourceFileAttribute.tag))
-                    return new SourceFileAttribute(cp, name, in);
-                else if (nameStr.equals(SyntheticAttribute.tag))
-                    return new SyntheticAttribute(cp, name, in);
-                else if (nameStr.equals(StackMap.tag))
-                    return new StackMap(cp, name, in);
-                else if (nameStr.equals(StackMapTable.tag))
-                    return new StackMapTable(cp, name, in);
-            }
-        }
+        if (first < 'E')
```

Comment:
The original code needs at most two comparison but the new code needs four in the worst case.

Reply:
Yes this is correct, the change was applied to improve readability and reduce the so called "What the?!?!" factor. I believe there were 2 instances of this same pattern.

We do sacrifice an additional check (in the worst case) but gain ease of understanding.